File: prompt_generator.py
#!/usr/bin/env python3
import random
from datetime import date, timedelta
from pathlib import Path


# Дата, соответствующая дню 1
START_DATE = date(2025, 2, 20)

# Папка с текстовыми файлами вариантов
SOURCES_DIR = Path(__file__).parent / "sources"


def load_variants(filename: str) -> list[str]:
    """Считывает варианты из текстового файла, по одному на строку."""
    path = SOURCES_DIR / filename
    with path.open("r", encoding="utf-8") as f:
        return [line.strip() for line in f if line.strip()]


def get_date_for_day(day_number: int) -> date:
    """Возвращает календарную дату для указанного порядкового номера."""
    if day_number < 1:
        raise ValueError("day_number должен быть >= 1")
    return START_DATE + timedelta(days=day_number - 1)


def choose_hair(gender: str) -> str:
    """Выбор варианта волосяного покрова в зависимости от пола."""
    if gender == "женский":
        variants = load_variants("hair_female.txt")
    else:
        variants = load_variants("hair_male.txt")
    return random.choice(variants)


def choose_palette(current_date: date) -> str:
    """
    Выбор цветовой палитры:
    - если воскресенье → всегда 'только красные тона'
    - иначе случайный вариант из palette.txt
    """
    # Monday=0 ... Sunday=6
    if current_date.weekday() == 6:
        return "только красные тона"

    variants = load_variants("palette.txt")
    return random.choice(variants)


def generate_prompt(day_number: int) -> str:
    """Генерирует текст промпта для указанного дня (1–1000)."""
    current_date = get_date_for_day(day_number)

    genders = load_variants("genders.txt")
    styles = load_variants("styles.txt")
    locations = load_variants("locations.txt")
    clothes = load_variants("clothes.txt")

    gender = random.choice(genders)
    style = random.choice(styles)
    location = random.choice(locations)
    clothes_choice = random.choice(clothes)
    hair = choose_hair(gender)
    palette = choose_palette(current_date)

    prompt = f"""
Перерисовать изображение в стиле: {style}
Текст: Сохранить арочный заголовок "MASTER'S TOUCH MEDITATION". Добавить под ним четкий подзаголовок "Day {day_number} of 1000". Текст и заголовок должен контрастировать с фоном.
Пол: {gender}
Локация: {location}
Одежда: {clothes_choice}
Волосяной покров: {hair}
Цветовая палитра: {palette}
Ориентация: Альбомная
""".strip()

    return prompt

if __name__ == "__main__":
    import sys

    # По умолчанию используем 308-й день, если аргумент не передан
    default_day = 308

    if len(sys.argv) > 1:
        try:
            day_number = int(sys.argv[1])
            if day_number < 1:
                raise ValueError
        except ValueError:
            print("Использование: python prompt_generator.py [day_number>=1]")
            sys.exit(1)
    else:
        day_number = default_day

    print(generate_prompt(day_number))

File: gas/prompt_generator.js
function loadVariantsFromSheet_(sheetName) {
  var ss = SpreadsheetApp.getActive();
  var sheet = ss.getSheetByName(sheetName);
  if (!sheet) {
    throw new Error('Sheet not found: ' + sheetName);
  }

  var lastRow = sheet.getLastRow();
  if (!lastRow) {
    throw new Error('Sheet has no data: ' + sheetName);
  }

  var values = sheet.getRange(1, 1, lastRow, 1).getValues();
  var variants = values
    .map(function (row) {
      return String(row[0]).trim();
    })
    .filter(function (value) {
      return value.length > 0;
    });

  if (!variants.length) {
    throw new Error('Sheet has no variants: ' + sheetName);
  }

  return variants;
}

function randomChoice_(values) {
  return values[Math.floor(Math.random() * values.length)];
}

function getDateForDay_(dayNumber) {
  if (dayNumber < 1) {
    throw new Error('dayNumber должен быть >= 1');
  }

  var baseDate = parseYmd_(BASE_DATE_YMD);
  var offsetMs = (dayNumber - 1) * 86400000;
  return new Date(baseDate.getTime() + offsetMs);
}

function chooseHair_(gender) {
  var sheetName = gender === 'женский' ? 'hair_female' : 'hair_male';
  return randomChoice_(loadVariantsFromSheet_(sheetName));
}

function choosePalette_(currentDate) {
  if (currentDate.getUTCDay() === 0) {
    return 'только красные тона';
  }

  return randomChoice_(loadVariantsFromSheet_('palette'));
}

function generatePrompt_(dayNumber) {
  var currentDate = getDateForDay_(dayNumber);

  var genders = loadVariantsFromSheet_('genders');
  var styles = loadVariantsFromSheet_('styles');
  var locations = loadVariantsFromSheet_('locations');
  var clothes = loadVariantsFromSheet_('clothes');

  var gender = randomChoice_(genders);
  var style = randomChoice_(styles);
  var location = randomChoice_(locations);
  var clothesChoice = randomChoice_(clothes);
  var hair = chooseHair_(gender);
  var palette = choosePalette_(currentDate);

  return [
    'Перерисовать изображение в стиле: ' + style,
    'Текст: Сохранить арочный заголовок "MASTER\'S TOUCH MEDITATION". ' +
      'Добавить под ним четкий подзаголовок "Day ' + dayNumber + ' of 1000". ' +
      'Текст и заголовок должен контрастировать с фоном.',
    'Пол: ' + gender,
    'Локация: ' + location,
    'Одежда: ' + clothesChoice,
    'Волосяной покров: ' + hair,
    'Цветовая палитра: ' + palette,
    'Ориентация: Альбомная'
  ].join('\n');
}
